### PR TITLE
SQL: Extract common parts into a separate QL project

### PR DESCRIPTION
**Pull request by costin**
First draft for extracting reusable components from SQL into their own project.
It is wip mainly to double check the tests

### CODE REVIEW

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/DateUtils.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static java.time.format.DateTimeFormatter.ISO_LOCAL_DATE;
+import static java.time.temporal.ChronoField.HOUR_OF_DAY;
+import static java.time.temporal.ChronoField.MILLI_OF_SECOND;
+import static java.time.temporal.ChronoField.MINUTE_OF_HOUR;
+import static java.time.temporal.ChronoField.NANO_OF_SECOND;
+import static java.time.temporal.ChronoField.SECOND_OF_MINUTE;
+
+public class DateUtils {
```
Comment (matriv) [inline x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/util/DateUtils.java]: Same here, regarding sharing.

**Review by matriv — COMMENTED** [review]
Left some comments, but looks good!

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/TopHitsAgg.java [hunk 2]
```diff
@@ -44,13 +44,13 @@ AggregationBuilder toBuilder() {
                 new FieldSortBuilder(sortField)
                     .order(sortOrder)
                     .missing(LAST.position())
-                    .unmappedType(sortFieldDataType.esType));
+                            .unmappedType(sortFieldDataType.esType()));
```
Comment (matriv) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/agg/TopHitsAgg.java]: identation issue

**Review by matriv — COMMENTED** [review]
Done with a few more comments.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/InProcessor.java [hunk 3]
```diff
@@ -41,7 +41,15 @@ public final void writeTo(StreamOutput out) throws IOException {
     @Override
     public Object process(Object input) {
         Object leftValue = processsors.get(processsors.size() - 1).process(input);
-        return apply(leftValue, Processors.process(processsors.subList(0, processsors.size() - 1), leftValue));
+        return apply(leftValue, process(processsors.subList(0, processsors.size() - 1), leftValue));
+    }
+
+    private static List<Object> process(List<Processor> processors, Object input) {
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/predicate/operator/comparison/InProcessor.java]: Yup. I got across it while trying to split generic processors vs SQL specific ones.

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/predicate/fulltext/FullTextUtilsTests.java [hunk 4]
```diff
@@ -18,7 +16,7 @@
 
 public class FullTextUtilsTests extends ESTestCase {
 
-    private final Source source = new Source(1, 1, StringUtil.EMPTY_STRING);
+    private final Source source = new Source(1, 1, "");
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/predicate/fulltext/FullTextUtilsTests.java]: For an empty string used in tests it seems excessive. Plus this is the first iteration, there's more polishing to do however I'd like to get the overall package structure done (such as splitting the `DataType` which is wired through statics and the type conversion).

**Review by astefan — APPROVED** [review]
Left some minor comments. LGTM.

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/OptionalArgument.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.ql.expression.function;
+
+/**
+ * Marker interface indicating that a function accepts one optional argument (typically the last one).
+ * This is used by the {@link FunctionRegistry} to perform validation of function declaration.
+ */
+public interface OptionalArgument {
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/OptionalArgument.java]: what do you mean? It's used by classes both in ql (`Iif`) and sql (`TopHits`).

File: x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/StringUtils.java [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+            .append(ISO_LOCAL_DATE)
+            .appendLiteral('T')
+            .appendValue(HOUR_OF_DAY, 2)
+            .appendLiteral(':')
+            .appendValue(MINUTE_OF_HOUR, 2)
+            .appendLiteral(':')
+            .appendValue(SECOND_OF_MINUTE, 2)
+            .appendFraction(MILLI_OF_SECOND, 3, 3, true)
+            .appendOffsetId()
+            .toFormatter(Locale.ROOT);
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/type/StringUtils.java]: Sure - it was copied from proto as such. I've fixed it in both places.

File: x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/AttributeMapTests.java [hunk 7]
```diff
@@ -3,10 +3,13 @@
...
  * you may not use this file except in compliance with the Elastic License.
  */
-package org.elasticsearch.xpack.sql.expression;
+package org.elasticsearch.xpack.ql.expression;
 
 import org.elasticsearch.test.ESTestCase;
-import org.elasticsearch.xpack.sql.tree.Source;
+import org.elasticsearch.xpack.ql.expression.Attribute;
+import org.elasticsearch.xpack.ql.expression.AttributeMap;
+import org.elasticsearch.xpack.ql.expression.UnresolvedAttribute;
```
[the remark this answers is not in the document]
Comment (costin) [reply x-pack/plugin/ql/src/test/java/org/elasticsearch/xpack/ql/expression/AttributeMapTests.java]: They aren't. Note the classes were just moved as they are without any import checks.
I'll probably do that in the branch to minimize further broad changes in the future.

**Comment by costin** [comment]
Superseeded by #50815

**Pull request closed without merging**